docs/03-code-internals/14-modifyclass.md

```diff
@@ -1,71 +1,138 @@
 id: modifyclass
-
 ---
-When using modifyClass in an initializer, you may see this warning in the console:
 
-> Attempted to modify "{name}", but it was already initialized earlier in the boot process (e.g. via a lookup()). Remove that lookup, or move the modifyClass call earlier in the boot process for changes to take effect.
+For advanced themes and plugins, Discourse offers the `modifyClass` system. This allows you to extend and override functionality in many of core's javascript classes.
 
-In theme/plugin development, there are two ways this error is normally introduced:
+## When to use `modifyClass`
+
+`modifyClass` should be a last resort, when your customization cannot be made via Discourse's more stable customization APIs (e.g. plugin-api methods, plugin outets, transformers).
+
+Core's code can change at any time. And therefore, customizations made via `modifyClass` could break at any time. When using this API, you should ensure that you have controls in place to catch those issues before they reach a production site. For example, you could add automated tests to the theme/plugin, or you could use a staging site to test incoming Discourse updates against your theme/plugin.
 
-### Adding a `lookup()` caused the error
+## Basic Usage
 
-If you `lookup()` a singleton too early in the boot process, it will cause any later `modifyClass` calls to fail. In this situation, you should try to move the lookup to happen later. For example, you would change something like this:
+`api.modifyClass` can be used to modify the functions and properties of any class which is accessible via the Ember resolver. That includes Discourse's routes, controllers, services and components.
 
+`modifyClass` takes two arguments:
+
+- `resolverName` (string) - construct this by using the type (e.g. component/controller/etc.), followed by a colon, followed by the (dasherized) filename name of the class. For example: `component:d-button`, `controller:user`, `route:application`, etc.
+
+- `callback` (function) - a function which receives the existing class definition, and then returns an extended version.
+
+For example, to modify the `click()` action on d-button:
+   
 ```js
-// Lookup service in initializer, then use it at runtime (bad!)
-export default apiInitializer("0.8", (api) => {
-  const composerService = api.container.lookup("service:composer");
-  api.composerBeforeSave(async () => {
-    composerService.doSomething();
-  });
-});
+api.modifyClass(
+  "component:d-button",
+  (Superclass) =>
+    class extends Superclass {
+      @action
+      click() {
+        console.log("button was clicked");
+        super.click();
+      }
+    }
+);
 ```
 
+The `class extends ...` syntax mimics that of [JS child classes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes#inheritance). In general, any syntax/features supported by child classes can be applied here. That includes `super`, static properties/functions, and more.
```

Comment:
We should include documentation on the side affects of omitting `super` when extending a class

Reply:
Can you expand on this? I'm not sure it's clear-cut - it really depends what you're doing. e.g. if you're extending a method like `didInsertElement()`, then you almost certainly want to call the core implementation first. But if you're aiming to change behaviour (e.g. overriding a `click()` action) then calling the core implementation might not make sense.